hypervisor/src/vm.rs

```diff
@@ -184,4 +215,15 @@ pub trait Vm: Send + Sync {
     fn state(&self) -> Result<VmState>;
     /// Set the VM state
     fn set_state(&self, state: &VmState) -> Result<()>;
+    /// set vmmops interface
+    fn set_vmmops(&mut self, vmmops: Arc<Mutex<dyn VmmOps>>) -> vm::Result<()>;
+}
+
+pub trait VmmOps: Send + Sync {
+    fn guest_mem_write(&self, buf: &[u8], gpa: u64) -> vm::Result<usize>;
+    fn guest_mem_read(&self, buf: &mut [u8], gpa: u64) -> vm::Result<usize>;
+    fn mmio_read(&self, addr: u64, data: &mut [u8]) -> vm::Result<()>;
+    fn mmio_write(&self, addr: u64, data: &[u8]) -> vm::Result<()>;
+    fn pio_read(&self, addr: u64, data: &mut [u8]) -> vm::Result<()>;
+    fn pio_write(&self, addr: u64, data: &[u8]) -> vm::Result<()>;
```

Comment:
@praveen-pk @liuw 
Sorry to comment a bit late on this PR, but here is how I think we could make this change less intrusive:
I would keep the `VmmOps` trait but I would remove both `guest_mem_write()` and `guest_mem_read()` from it. Instead, we could expect the `hypervisor` to rely on `vm-memory`, which means we could set/pass a `GuestMemoryAtomic<GuestMemoryMmap>`, therefore the `hypervisor` crate could directly deal with read/write from/to the guest memory.
This would simplify the `VmmOps` trait, but the big benefit of this approach would be to implement the `VmmOps` from the `DeviceManager` directly instead of `Vm`. I'm really not a huge fan of having the entire `Vm` object being locked every time an operation from the `VmmOps` is performed.

@rbradford @sameo WDYT?

Reply:
@sboeuf Since the hypervisor crate already depends on the `vm-memory` crate, that would make sense.
What should be passed to the hypervisor crate is something more generic than `GuestMemoryAtomic<GuestMemoryMmap>` though. I don't think we should expect the hypervisor crate could modify the guest address space and could only pass something that implements `GuestAddressSpace`?

Reply:
I think if we don't pass a `GuestMemoryAtomic<>`, we might run into the issue that some memory regions might have been added, hence the hypervisor would not have access to the updated `GuestMemoryMmap`.
And the only method from the `GuestAddressSpace` trait is `memory()`, which is defined as follows:
```
    /// Return an object (e.g. a reference or guard) that can be used
    /// to access memory through this address space.  The object provides
    /// a consistent snapshot of the memory map.
    fn memory(&self) -> Self::T;
```
So the problem is that it simply provides a snapshot at a given time, but I'm worried we might break things for memory hotplug.

Reply:
If we don't agree implementing vmmops in the Vm we have to rewrite the patches almost completely. It would be better if we have a consensus ASAP.

Reply:
> I think if we don't pass a `GuestMemoryAtomic<>`, we might run into the issue that some memory regions might have been added, hence the hypervisor would not have access to the updated `GuestMemoryMmap`.
> And the only method from the `GuestAddressSpace` trait is `memory()`, which is defined as follows:
> 
> ```
>     /// Return an object (e.g. a reference or guard) that can be used
>     /// to access memory through this address space.  The object provides
>     /// a consistent snapshot of the memory map.
>     fn memory(&self) -> Self::T;
> ```
> 
> So the problem is that it simply provides a snapshot at a given time, but I'm worried we might break things for memory hotplug.

So we're forcing the VMM to use an hotpluggable memory implementation, which is not something that's portable. Then I think it makes sense to keep those as part of the VmmOps trait and let the VMM handle memory operations depending on how it decided to implement its guest address space.

Reply:
Well I understand your frustration, but it's important that we merge the right design the first time, otherwise this will cost everybody more time if we decide to modify it later. Moreover, I actually think the proposal would simplify the PR.
@rbradford WDYT?

Reply:
@sameo 
>So we're forcing the VMM to use an hotpluggable memory implementation, which is not something that's portable.

Ok that's fair, even if I think `GuestMemoryAtomic<>` should be the smallest unit, even for VMM without hotplug needs.

I would still like to avoid locking the entire `Vm` structure. So how about we define a new structure which implements `VmmOps` and which has a reference to both `MemoryManager` and `DeviceManager`? This structure does not need to be shared between the `vmm` and `hypervisor` crate, therefore it would greatly simplify the whole `Mutex` complexity.
Something like
```rust
struct VmOps {
    memory_manager: ...
    device_manager: ...
}
impl VmmOps for VmOps {
...
}
```

Reply:
> > If we don't agree implementing vmmops in the Vm, we have to rewrite the patches almost completely. It would be better if we have an agreement ASAP.
> 
> Well I understand your frustration, but it's important that we merge the right design the first time, otherwise this will cost everybody more time if we decide to modify it later. Moreover, I actually think the proposal would simplify the PR.
> @rbradford WDYT?

I can't agree more. I am not frustrated thought. :)

Reply:
@russell-islam we talked about this offline with @sameo and @rbradford. I'm gonna have a quick try at implementing the `VmOps` solution mentioned above. I'll get back to you with when ready.

Reply:
> @russell-islam we talked about this offline with @sameo and @rbradford. I'm gonna have a quick try at implementing the `VmOps` solution mentioned above. I'll get back to you when ready.

Sounds good. :+1:

Reply:
@russell-islam here is the simplified version I was thinking about: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1793
I'm just missing a few things regardind the VM exits (moving things from `vmm/src/cpu.rs` to `hypervisor/src/kvm/mod.rs`) but the point is to show how I got rid of the locks. Let me know what you think about it.

Reply:
/cc @sameo @rbradford